File: backend/app/schemas/auth_extended.py
"""
Extended Auth Schemas for Roles, Menu Permissions, and Keycloak
"""
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, Field


# ============ Role Schemas ============

class RoleBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=50, description="ชื่อ unique ของ role")
    display_name: str = Field(..., min_length=1, max_length=100, description="ชื่อแสดงผล")
    description: Optional[str] = Field(None, description="คำอธิบาย")
    level: int = Field(0, ge=0, le=100, description="ระดับสิทธิ์ (0-100)")
    is_active: bool = True


class RoleCreate(RoleBase):
    pass


class RoleUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=50)
    display_name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    level: Optional[int] = Field(None, ge=0, le=100)
    is_active: Optional[bool] = None


class RoleResponse(RoleBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


# ============ Menu Item Schemas ============

class MenuItemBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100, description="ชื่อ unique ของเมนู")
    display_name: str = Field(..., min_length=1, max_length=100, description="ชื่อแสดงผล")
    path: str = Field(..., description="URL path")
    icon: Optional[str] = Field(None, max_length=50, description="ชื่อ icon")
    parent_id: Optional[int] = Field(None, description="Parent menu ID")
    order: int = Field(0, ge=0, description="ลำดับการแสดงผล")
    is_visible: bool = True
    description: Optional[str] = None


class MenuItemCreate(MenuItemBase):
    pass


class MenuItemUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    display_name: Optional[str] = Field(None, min_length=1, max_length=100)
    path: Optional[str] = None
    icon: Optional[str] = None
    parent_id: Optional[int] = None
    order: Optional[int] = None
    is_visible: Optional[bool] = None
    description: Optional[str] = None


class MenuItemResponse(MenuItemBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # Note: children is not included to avoid lazy loading issues with async

    class Config:
        from_attributes = True


# ============ Role Menu Permission Schemas ============

class RoleMenuPermissionBase(BaseModel):
    role_id: int
    menu_item_id: int
    can_view: bool = True
    can_edit: bool = False
    can_delete: bool = False


class RoleMenuPermissionCreate(RoleMenuPermissionBase):
    pass


class RoleMenuPermissionUpdate(BaseModel):
    can_view: Optional[bool] = None
    can_edit: Optional[bool] = None
    can_delete: Optional[bool] = None


class RoleMenuPermissionResponse(RoleMenuPermissionBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PermissionMatrixItem(BaseModel):
    """Item for permission matrix view"""
    menu_item_id: int
    menu_name: str
    menu_display_name: str
    menu_path: str
    can_view: bool = False
    can_edit: bool = False
    can_delete: bool = False


class RolePermissionMatrix(BaseModel):
    """Complete permission matrix for a role"""
    role_id: int
    role_name: str
    role_display_name: str
    permissions: List[PermissionMatrixItem]


class BulkPermissionUpdate(BaseModel):
    """Bulk update permissions for a role"""
    role_id: int
    permissions: List[dict]  # [{menu_item_id: int, can_view: bool, can_edit: bool, can_delete: bool}]


# ============ Keycloak Config Schemas ============

class AllowedUser(BaseModel):
    """User ที่อนุญาตให้เข้าสู่ระบบผ่าน Keycloak"""
    username: str
    role: str = "viewer"  # Default role เป็น viewer


class KeycloakConfigBase(BaseModel):
    is_enabled: bool = False
    server_url: Optional[str] = Field(None, max_length=500)
    realm: Optional[str] = Field(None, max_length=100)
    client_id: Optional[str] = Field(None, max_length=100)
    redirect_uri: Optional[str] = Field(None, max_length=500)
    scope: str = "openid profile email"
    
    # Role mapping
    admin_role: str = "admin"
    editor_role: str = "editor"
    viewer_role: str = "viewer"
    default_role: str = "viewer"
    
    # Options
    auto_create_user: bool = True
    sync_user_info: bool = True
    
    # Allowed users
    allowed_users: List[AllowedUser] = Field(default_factory=list)


class KeycloakConfigCreate(KeycloakConfigBase):
    client_secret: Optional[str] = Field(None, max_length=500)


class KeycloakConfigUpdate(BaseModel):
    is_enabled: Optional[bool] = None
    server_url: Optional[str] = None
    realm: Optional[str] = None
    client_id: Optional[str] = None
    client_secret: Optional[str] = None
    redirect_uri: Optional[str] = None
    scope: Optional[str] = None
    admin_role: Optional[str] = None
    editor_role: Optional[str] = None
    viewer_role: Optional[str] = None
    default_role: Optional[str] = None
    auto_create_user: Optional[bool] = None
    sync_user_info: Optional[bool] = None
    allowed_users: Optional[List[AllowedUser]] = None


class KeycloakConfigResponse(KeycloakConfigBase):
    id: int
    updated_by: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    # Note: client_secret is not included in response for security

    class Config:
        from_attributes = True


class KeycloakConfigPublic(BaseModel):
    """Public config for frontend login button"""
    is_enabled: bool
    server_url: Optional[str] = None
    realm: Optional[str] = None
    client_id: Optional[str] = None
    redirect_uri: Optional[str] = None
    scope: Optional[str] = None


class KeycloakLoginResponse(BaseModel):
    """Response for Keycloak login initiation"""
    auth_url: str
    state: str


class KeycloakCallbackRequest(BaseModel):
    """Request for Keycloak callback"""
    code: str
    state: str


class KeycloakUserInfo(BaseModel):
    """User info from Keycloak"""
    sub: str  # Subject (user ID in Keycloak)
    preferred_username: Optional[str] = None
    email: Optional[str] = None
    name: Optional[str] = None
    given_name: Optional[str] = None
    family_name: Optional[str] = None
    roles: List[str] = []


# ============ Keycloak User Mapping Schemas ============

class KeycloakUserMappingBase(BaseModel):
    """Base schema for Keycloak user mapping"""
    keycloak_user_id: str = Field(..., description="Keycloak user ID")
    keycloak_username: str = Field(..., description="Keycloak username")
    keycloak_email: Optional[str] = None
    keycloak_full_name: Optional[str] = None
    local_role: str = Field(..., description="Local role (admin, editor, viewer)")
    is_enabled: bool = True
    user_attributes: Optional[dict] = None


class KeycloakUserMappingCreate(KeycloakUserMappingBase):
    """Schema for creating user mapping"""
    pass


class KeycloakUserMappingUpdate(BaseModel):
    """Schema for updating user mapping"""
    local_role: Optional[str] = None
    is_enabled: Optional[bool] = None
    user_attributes: Optional[dict] = None


class KeycloakUserMappingResponse(KeycloakUserMappingBase):
    """Schema for user mapping response"""
    id: int
    created_by: Optional[str] = None
    updated_by: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class KeycloakUserFromServer(BaseModel):
    """User data from Keycloak server"""
    id: str
    username: str
    email: Optional[str] = None
    firstName: Optional[str] = None
    lastName: Optional[str] = None
    enabled: bool = True
    emailVerified: Optional[bool] = None
    createdTimestamp: Optional[int] = None
    attributes: Optional[dict] = None


class KeycloakUserWithMapping(BaseModel):
    """Keycloak user with local role mapping"""
    keycloak_user: KeycloakUserFromServer
    mapping: Optional[KeycloakUserMappingResponse] = None
    has_mapping: bool = False

File: frontend/src/services/authExtended.ts
/**
 * Auth Extended API Services
 * API services for roles, menu permissions, Keycloak SSO, and user mapping
 */
import { apiGet, apiPost, apiPut, apiDelete } from '../lib/api';

// ============ Types ============

export interface Role {
  id: number;
  name: string;
  display_name: string;
  description?: string;
  level: number;
  is_active: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface RoleCreate {
  name: string;
  display_name: string;
  description?: string;
  level?: number;
  is_active?: boolean;
}

export interface RoleUpdate {
  name?: string;
  display_name?: string;
  description?: string;
  level?: number;
  is_active?: boolean;
}

export interface MenuItem {
  id: number;
  name: string;
  display_name: string;
  path: string;
  icon?: string;
  parent_id?: number;
  order: number;
  is_visible: boolean;
  description?: string;
  created_at?: string;
  updated_at?: string;
  children?: MenuItem[];
}

export interface MenuItemCreate {
  name: string;
  display_name: string;
  path: string;
  icon?: string;
  parent_id?: number;
  order?: number;
  is_visible?: boolean;
  description?: string;
}

export interface RoleMenuPermission {
  id: number;
  role_id: number;
  menu_item_id: number;
  can_view: boolean;
  can_edit: boolean;
  can_delete: boolean;
  created_at?: string;
  updated_at?: string;
}

export interface PermissionMatrixItem {
  menu_item_id: number;
  menu_name: string;
  menu_display_name: string;
  menu_path: string;
  can_view: boolean;
  can_edit: boolean;
  can_delete: boolean;
}

export interface RolePermissionMatrix {
  role_id: number;
  role_name: string;
  role_display_name: string;
  permissions: PermissionMatrixItem[];
}

export interface BulkPermissionUpdate {
  role_id: number;
  permissions: {
    menu_item_id: number;
    can_view?: boolean;
    can_edit?: boolean;
    can_delete?: boolean;
  }[];
}

export interface AllowedUser {
  username: string;
  role: string;
}

export interface KeycloakConfig {
  id: number;
  is_enabled: boolean;
  server_url: string;
  realm: string;
  client_id: string;
  client_secret: string;
  redirect_uri: string;
  scope?: string;
  admin_role?: string;
  editor_role?: string;
  viewer_role?: string;
  default_role?: string;
  auto_create_user?: boolean;
  sync_user_info?: boolean;
  allowed_users?: AllowedUser[];
  updated_by?: string;
  created_at?: string;
  updated_at?: string;
}

export interface KeycloakConfigCreate {
  is_enabled: boolean;
  server_url: string;
  realm: string;
  client_id: string;
  client_secret: string;
  redirect_uri: string;
  scope?: string;
  admin_role?: string;
  editor_role?: string;
  viewer_role?: string;
  default_role?: string;
  auto_create_user?: boolean;
  sync_user_info?: boolean;
  allowed_users?: AllowedUser[];
}

export interface KeycloakConfigUpdate {
  is_enabled?: boolean;
  server_url?: string;
  realm?: string;
  client_id?: string;
  client_secret?: string;
  redirect_uri?: string;
  scope?: string;
  admin_role?: string;
  editor_role?: string;
  viewer_role?: string;
  default_role?: string;
  auto_create_user?: boolean;
  sync_user_info?: boolean;
  allowed_users?: AllowedUser[];
}

export interface KeycloakPublicConfig {
  is_enabled: boolean;
  server_url?: string;
  realm?: string;
  client_id?: string;
  redirect_uri?: string;
  scope?: string;
}

export interface KeycloakLoginResponse {
  auth_url: string;
  state: string;
}

// Keycloak User Management Types
export interface KeycloakUserFromServer {
  id: string;
  username: string;
  email?: string;
  firstName?: string;
  lastName?: string;
  enabled: boolean;
  emailVerified?: boolean;
  createdTimestamp?: number;
  attributes?: Record<string, any>;
}

export interface KeycloakUserMapping {
  id: number;
  keycloak_user_id: string;
  keycloak_username: string;
  keycloak_email?: string;
  keycloak_full_name?: string;
  local_role: string;
  is_enabled: boolean;
  user_attributes?: Record<string, any>;
  created_by?: string;
  updated_by?: string;
  created_at?: string;
  updated_at?: string;
}

export interface KeycloakUserMappingCreate {
  keycloak_user_id: string;
  keycloak_username: string;
  keycloak_email?: string;
  keycloak_full_name?: string;
  local_role: string;
  is_enabled?: boolean;
  user_attributes?: Record<string, any>;
}

export interface KeycloakUserMappingUpdate {
  local_role?: string;
  is_enabled?: boolean;
  user_attributes?: Record<string, any>;
}

export interface KeycloakUserWithMapping {
  keycloak_user: KeycloakUserFromServer;
  mapping?: KeycloakUserMapping;
  has_mapping: boolean;
}

// ============ Roles API ============

export const rolesApi = {
  list: async (params?: { is_active?: boolean }) => {
    return apiGet<Role[]>('/roles/', params);
  },
  
  get: async (id: number) => {
    return apiGet<Role>(`/roles/${id}/`);
  },
  
  create: async (data: RoleCreate) => {
    return apiPost<Role>('/roles/', data);
  },
  
  update: async (id: number, data: RoleUpdate) => {
    return apiPut<Role>(`/roles/${id}/`, data);
  },
  
  delete: async (id: number) => {
    return apiDelete(`/roles/${id}/`);
  },
  
  initDefaults: async () => {
    return apiPost<Role[]>('/roles/init-defaults/', {});
  }
};

// ============ Menu Items API ============

export const menuItemsApi = {
  list: async (params?: { is_visible?: boolean; parent_id?: number }) => {
    return apiGet<MenuItem[]>('/menu-items/', params);
  },
  
  listAll: async () => {
    return apiGet<MenuItem[]>('/menu-items/all');
  },
  
  get: async (id: number) => {
    return apiGet<MenuItem>(`/menu-items/${id}`);
  },
  
  create: async (data: MenuItemCreate) => {
    return apiPost<MenuItem>('/menu-items/', data);
  },
  
  update: async (id: number, data: Partial<MenuItemCreate>) => {
    return apiPut<MenuItem>(`/menu-items/${id}`, data);
  },
  
  delete: async (id: number) => {
    return apiDelete(`/menu-items/${id}`);
  },
  
  initDefaults: async () => {
    return apiPost<MenuItem[]>('/menu-items/init-defaults', {});
  }
};

// ============ Permissions API ============

export const permissionsApi = {
  list: async (params?: { role_id?: number; menu_item_id?: number }) => {
    return apiGet<RoleMenuPermission[]>('/permissions/', params);
  },
  
  getMatrix: async () => {
    return apiGet<RolePermissionMatrix[]>('/permissions/matrix');
  },
  
  getRolePermissions: async (roleId: number) => {
    return apiGet<RolePermissionMatrix>(`/permissions/role/${roleId}`);
  },
  
  getCurrentUserPermissions: async () => {
    return apiGet<PermissionMatrixItem[]>('/permissions/user/current');
  },
  
  create: async (data: Omit<RoleMenuPermission, 'id' | 'created_at' | 'updated_at'>) => {
    return apiPost<RoleMenuPermission>('/permissions/', data);
  },
  
  update: async (id: number, data: { can_view?: boolean; can_edit?: boolean; can_delete?: boolean }) => {
    return apiPut<RoleMenuPermission>(`/permissions/${id}`, data);
  },
  
  delete: async (id: number) => {
    return apiDelete(`/permissions/${id}`);
  },
  
  bulkUpdate: async (data: BulkPermissionUpdate) => {
    return apiPost<RoleMenuPermission[]>('/permissions/bulk-update', data);
  },
  
  initDefaults: async () => {
    return apiPost<RoleMenuPermission[]>('/permissions/init-defaults', {});
  }
};

// ============ Keycloak API ============

export const keycloakApi = {
  // Config management
  getConfig: async () => {
    return apiGet<KeycloakConfig>('/keycloak/config/');
  },
  
  getPublicConfig: async () => {
    return apiGet<KeycloakPublicConfig>('/keycloak/public-config');
  },
  
  createConfig: async (data: KeycloakConfigCreate) => {
    return apiPost<KeycloakConfig>('/keycloak/config/', data);
  },
  
  updateConfig: async (data: Partial<KeycloakConfigCreate>) => {
    return apiPut<KeycloakConfig>('/keycloak/config/', data);
  },
  
  deleteConfig: async () => {
    return apiDelete('/keycloak/config/');
  },
  
  testConnection: async () => {
    return apiPost<{ success: boolean; message: string; realm_info?: any; error?: string }>('/keycloak/test-connection/', {});
  },
  
  // Authentication flow
  initiateLogin: async () => {
    return apiGet<KeycloakLoginResponse>('/keycloak/login');
  },
  
  handleCallback: async (code: string, state: string) => {
    return apiPost<{
      access_token: string;
      token_type: string;
      user: {
        id: number;
        username: string;
        full_name: string;
        email?: string;
        role: string;
      };
    }>('/keycloak/callback', { code, state });
  },
  
  // User management
  getUsers: async (params?: { search?: string; max_results?: number }) => {
    return apiGet<KeycloakUserFromServer[]>('/keycloak/users', params);
  },
  
  getUsersWithMappings: async (params?: { search?: string; max_results?: number }) => {
    return apiGet<KeycloakUserWithMapping[]>('/keycloak/users-with-mappings', params);
  },
  
  // User mappings
  getUserMappings: async () => {
    return apiGet<KeycloakUserMapping[]>('/keycloak/user-mappings');
  },
  
  createUserMapping: async (data: KeycloakUserMappingCreate) => {
    return apiPost<KeycloakUserMapping>('/keycloak/user-mappings', data);
  },
  
  updateUserMapping: async (mappingId: number, data: KeycloakUserMappingUpdate) => {
    return apiPut<KeycloakUserMapping>(`/keycloak/user-mappings/${mappingId}`, data);
  },
  
  deleteUserMapping: async (mappingId: number) => {
    return apiDelete(`/keycloak/user-mappings/${mappingId}`);
  },
  
  getUserMappingByKeycloakId: async (keycloakUserId: string) => {
    return apiGet<KeycloakUserMapping>(`/keycloak/user-mappings/by-keycloak-id/${keycloakUserId}`);
  }
};

